### DISCUSSION TITLE
How to consistently handle "Booked" and "Pending" with GoCardless

### DISCUSSION THREAD

**Discussion by tboby** [firefly-iii/firefly-iii#8678] [Firefly III]
With the recent importer updates it looks like transactions now come tagged as "pending" or "booked", which is great!

However, I've started to notice duplicate transactions, as well as pending but never completed transactions. Sometimes the pending transaction is dated before the booked, sometimes after. I don't think there is always both a pending and booked. The booked sometimes has more details than pending.

My understanding is that the importer checks deleted transactions for duplicates, so I can't simply delete all pending.

How does one simultaneously:
1. Delete "Pending" transactions which never become booked
2. Avoid the importer failing to import "Booked" transactions because they're a duplicate

**Comment by JC5** [comment]
If I'm not mistaken, booked and pending transactions will never be each others duplicate.

**Reply by chris-redbeed** [reply]
But should there? Or do an Update on the "pending" on if you find a "booked" transaction? 
Currently I have the problem that I have most of the time a transaction twice in my history.

**Reply by JC5** [reply]
> But should there? Or do an Update on the "pending" on if you find a "booked" transaction?

Based on what, exactly? The problem is that all the details can change when a transaction goes from pending to booked. I can't match the first on the second.

**Reply by chris-redbeed** [reply]
Oh, I see—I hadn't considered that, but you're right; only the price is the same. Perhaps an option to exclude pending transactions and import only booked ones would be beneficial?

Pending transactions shouldn't have a direct impact on your balance yet, should they?

-> Therefore, introducing an option to instruct the importer to exclude transactions with a "pending" status could be a solution.

**Reply by JC5** [reply]
Create a rule that deletes transactions with the "pending" tag.

**Reply by chris-redbeed** [reply]
Good point, was thinking more about a solution for everyone that uses the GoCardless.
Maybe if I find the time I suggest a Pull request - but for know a rule would be fine you right.

**Comment by EricVanCaenenberghe** [comment]
Hi, if you based your "detect duplicates" on the external identifier then they are not considered as duplicates (probably depending on the bank). 

I implemented a rule to discard pending transactions from being enriched (destination account, category, link to contract) except concerning a budget (pending budget of 1 cent). I would like to find a method to clean the pending transaction if the booked version arrives (I see them in the correct order, because my bank eliminates the pending transactions once there is a booked version of it).

valid for my bank:
pending = extid1 notes section without end2endReference (middle of the day)
booked v1 = extid2 notes section identical to the pending except it starts with an end2endReference (middle of the day: 1-2 hours after the pending transaction)
booked v2 = extid2 notes section identical to the pending except it starts with StatementReference + end2endReference (only delivered on a working business day) 
<= the last one is disregarded because it is a duplicate based on the external identifier
(it is a pitty to loose the StatementReference because that reference is handy when reconcialing an account and you think there is a gap)

Little balloons:
- duplicate detection based on external identifier and notes section: if external identifier is identical but notes section not then copy notes section 
- automatic deletion of pending transactions when 90% of the notes section matches with a newly arrived booked transaction?

greetings
